final-project/static/scripts.js

```diff
@@ -220,207 +233,386 @@ function displayProduct(jsonOfProductInfo) {
+  var editRequestsFromUser = {};
+
+  // Find any requested edits from the user.
+  for (var i = 0; i < inputsAvailableToUser.length; i++) {
+    // If the input has a value, save the information.
+    if (inputsAvailableToUser[i].value && inputsAvailableToUser[i].id != 'save-edits') {
+
+      // Identify which information the user would like to change.
+      var infoToChange = inputsAvailableToUser[i].parentNode.name;
+
+      // If it is not a reference, then store the info.
+      if (!infoToChange.startsWith("reference-form")) {
+        // Create a key. Set it equal to the form name.
+        var itemToEdit = inputsAvailableToUser[i].parentNode.name;
+        // Create a value. Store the users edits here.
+        var requestedUpdateFromUser = inputsAvailableToUser[i].value;
+
+        // Store this key-value pair into a single javascript object.
+        // Then we can make a single request to our server.
+        editRequestsFromUser[itemToEdit] = requestedUpdateFromUser;
+      }
+      // If the requested edit was for a reference,
+      // then make an additional array to organize the data.
+      else {
+        // Make a javascript array to store our reference data into.
+        // This will be the 'value' which corresponds to the reference 'key'.
+        var editToThisReference = {};
+
+        // If this edit is for a reference's title, save the title.
+        if (inputsAvailableToUser[i].name.startsWith('reference-title')) {
+
+          // Save the title edit.
+          var referenceTitle = inputsAvailableToUser[i].value;
+
+          // Store this information in the array for this reference id.
+          editToThisReference['referenceTitle'] = referenceTitle;
+
+          // Check for requested changes to the link.
+          if (inputsAvailableToUser[i+1].value) {
+
+            // If there is a change to the link then save the change.
+            var referenceLink = inputsAvailableToUser[i+1].value;
+
+            // Store this information in the array for this reference id.
+            editToThisReference['referenceLink'] = referenceLink;
+
+            // Otherwise, do not save a link value.
+          }
+
+          // No need to check the link input item again.
+          // Increment i to jump to the next relevant input.
+          i++;
+        }
+        // Otherwise, if the requested edit corresponds to a reference's link,
+        // only save the link.
+        else {
+
+          // Save the link.
+          var referenceLink = inputsAvailableToUser[i].value;
+          // There is no change to this reference's title.
+
+          // Store this information in the array for this reference id.
+          editToThisReference['referenceLink'] = referenceLink;
+        }
+
+        // Find the reference id.
+        // This will be how we determine which reference to update in our db.
+        var referenceId = inputsAvailableToUser[i].parentNode.parentNode.id;
```

Comment:
For instance, you are traversing the nodes directly. That seems like a lot for what you are trying to do.
It's easier to set ids on the elements you care about and use findElementById rather than depending on the HTML structure, which may easily change in a template refactor.